### DISCUSSION TITLE
Assincronia

### DISCUSSION THREAD

**Discussion by ThiagoBussola** [TI-UNICESUMAR/2024-desafio-profissional-v-ESOFT5S-A#2] [Node.js]
O paradigma assíncrono é uma parte crucial do desenvolvimento Javascript, entender suas peculiaridades irá auxiliar você a criar aplicar melhores práticas para desenvolver e performar suas aplicações.

Para auxiliar seus conhecimentos nesse tema, leia o capítulo 3 do livro: Node.js Design Patterns 3 edition.
Coloque neste tópico suas dúvidas, pontos que achou interessantes e materiais complementares que você utilizou para entender o tema (artigos, vídeos, podcasts, etc).

**Comment by kaikdorvalo** [comment]
Algo que achei curioso a primeira vez que fiz uma api com Spring Boot no java, é que não havia necessidade de async e await nas requisições, pois sempre utilizei o nodejs. Estudando mais a fundo, entendi como funciona o tratamento de concorrência e como isso era tratado de forma multithreading no Spring e Single Thread no Nodejs, por isso havia a necessidade de se utilizar async e await, justamente para não bloquear a execução do código até que a Promise fosse resolvida.

Como o professor comentou em aula, temos requisições que necessitam de await e outras que não precisam. Um exemplo de quando precisa de await, é quando estamos dentro de funções assíncronas e, fazemos a busca de um usuário no banco, por exemplo. Logo em seguida queremos realizar algo com esse usuário, como extrair o seu nome, data de nascimento, etc.

Quando estamos dentro de funções síncronas, podemos utilizar o .then() para lidar com as promessas, excluindo o uso do await.

Podemos perceber também que, quando temos uma sequência de chamadas assíncronas, uma logo depois da outra, a primeira será disparada, a segunda somente quando a primeira for resolvida, a terceira depois da segunda ser resolvida e por assim vai. Nesse caso, se nenhuma promise depende da outra para ser chamada, pode ser utilizado o Promises.all para que todas sejam disparadas ao mesmo tempo,  levando a um grande ganho de desempenho.

**Reply by willyantomaz** [reply]
Percebi a mesma coisa, Kaik. Ano passado, quando começamos a estudar POO, nem utilizamos o async e o await. Porém, agora no trabalho, usando Flutter para o front-end e o Java Spring para a minha API, tive que aprender a utilizar o FUTURE, async e await nas minhas requisições.

**Comment by joaoagr1** [comment]
Exemplo de uso de assincronia conforme o professor tinha comentado em sala, porém de forma implicita com uma requisição do tipo fetch que retorna uma Promise. Isso significa que ela trabalha de forma assíncrona e permite que você trate a resposta da requisição HTTP quando estiver pronta:

```
`function fetchAddSpell(){
    characterId = localStorage.getItem('idCharacter');
    nomeSpell = document.getElementById('SpellNamePost').value;
    descriptionSpell = document.getElementById('descriptionSpellPost').value;


    let url = 'http://localhost:8080/spells';
    let data = {
        character_id: characterId,  
        name: nomeSpell, 
        description: descriptionSpell
    };

    fetch(url, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
    })
    .then(response => response.json())
    .then(data => console.log(data))
    .catch((error) => {
        console.error('Error:', error);
    });`
```

**Comment by EnzoLavieri** [comment]
Alguns pontos que eu achei interessantes nesse capítulo do livro são:

- **O conceito de callback**: Um callback é uma função que é passada como argumento para outra função, e que é invocada quando essa função termina sua execução. Os callbacks são a forma mais básica de lidar com operações assíncronas no Node.js, mas também podem levar a problemas como o "callback hell", que é quando temos muitos callbacks aninhados, tornando o código difícil de ler e manter.

- **O padrão Promise**: Uma promise é um objeto que representa o resultado futuro de uma operação assíncrona. As promises permitem que escrevamos código assíncrono de forma mais declarativa e encadeada, usando os métodos then e catch para tratar os casos de sucesso e erro. As promises também podem ser combinadas usando métodos estáticos como Promise.all, Promise.race, Promise.resolve e Promise.reject.

- **O uso de async e await**: O async e o await são recursos da linguagem Javascript que permitem que escrevamos código assíncrono de forma mais simples e intuitiva, usando a palavra-chave await para esperar o resultado de uma promise, e a palavra-chave async para marcar uma função que retorna uma promise. O uso de async e await torna o código mais legível e fácil de depurar, mas também requer alguns cuidados, como o tratamento de erros com try e catch.
